parse/planner/planner.go

```diff
@@ -0,0 +1,2392 @@
+	}
+
+	if node.IsTable {
+		return nil, fmt.Errorf("procedure returns a table, not a scalar value")
+	}
+
+	if len(node.Fields) != 1 {
+		return nil, fmt.Errorf("procedures in expressions must return exactly one value, received %d", len(node.Fields))
+	}
+
+	return node.Fields[0].Type.Copy(), nil
+}
+
+// anonField creates an anonymous field with the given data type.
+func anonField(dt *types.DataType) *Field {
+	return &Field{
+		val: dt,
+	}
+}
+
+// table takes a parse.Table interface and returns the plan and relation
+// for the table.
+func (s *scopeContext) table(node parse.Table) (*Scan, *Relation, error) {
+	switch node := node.(type) {
+	default:
+		panic(fmt.Sprintf("unexpected parse table type %T", node))
+	case *parse.RelationTable:
+		// either a CTE or a physical table
+		alias := node.Table
+		if node.Alias != "" {
+			alias = node.Alias
+		}
+
+		var scanTblType TableSourceType
+		var rel *Relation
+		if physicalTbl, ok := s.plan.Schema.FindTable(node.Table); ok {
+			scanTblType = TableSourcePhysical
+			rel = relationFromTable(physicalTbl)
+		} else if cte, ok := s.plan.CTEs[node.Table]; ok {
+			scanTblType = TableSourceCTE
+			rel = cte
+		} else {
+			return nil, nil, fmt.Errorf(`unknown table "%s"`, node.Table)
+		}
+
+		for _, col := range rel.Fields {
+			col.Parent = alias
+		}
+
+		return &Scan{
+			Source: &TableScanSource{
+				TableName: node.Table,
+				Type:      scanTblType,
+				rel:       rel.Copy(),
+			},
+			RelationName: alias,
+		}, rel, nil
+	case *parse.RelationSubquery:
+		if node.Alias == "" {
+			return nil, nil, fmt.Errorf("join against subquery must have an alias")
+		}
+
+		// we pass an empty relation because the subquery can't
+		// refer to the current relation, but they can be correlated against some
+		// outer relation.
+		// for example, "select * from users u inner join (select * from posts where posts.id = u.id) as p on u.id=p.id;"
+		// is invalid, but
+		// "select * from users where id = (select posts.id from posts inner join (select * from posts where id = users.id) as s on s.id=posts.id);"
+		// is valid
+		subq, rel, err := s.planSubquery(node.Subquery, &Relation{})
+		if err != nil {
+			return nil, nil, err
+		}
+
+		for _, col := range rel.Fields {
+			col.Parent = node.Alias
+		}
+
+		return &Scan{
+			Source:       subq,
+			RelationName: node.Alias,
+		}, rel, nil
+	case *parse.RelationFunctionCall:
+		if node.Alias == "" {
+			return nil, nil, fmt.Errorf("join against procedure calls must have an alias")
+		}
+
+		// the function call must either be a procedure or foreign procedure that returns
+		// a table.
+
+		var args []LogicalExpr
+		var contextArgs []LogicalExpr
+		var procReturns *types.ProcedureReturn
+		var isForeign bool
+		if proc, ok := s.plan.Schema.FindProcedure(node.FunctionCall.FunctionName()); ok {
+			procReturns = proc.Returns
+
+			procCall, ok := node.FunctionCall.(*parse.ExpressionFunctionCall)
+			if !ok {
```

Comment:
No, this is correct. Procedure calls are parsed as `parse.ExpressionFunctionCall`, since syntactically, they are no different than a normal function:

```
$a := abs(-1);
// looks the same to the parser as
$b := my_specific_procedure(-1);
```

Reply:
ok now I get it